src/transformStateWithClones.js

```diff
@@ -7,7 +7,32 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const array = [];
+  const newState = { ...state };
+
+  for (const { type, ...rest } of actions) {
+    if (type === 'addProperties') {
+      Object.assign(newState, rest.extraData);
+    }
+
+    if (type === 'removeProperties') {
+      for (const key of rest.keysToRemove) {
+        delete newState[key];
+      }
+    }
+
+    if (type === 'clear') {
+      for (const key in newState) {
+        delete newState[key];
+      }
+    }
+
+    const newState1 = { ...newState };
```

Comment:
Naming issue: The variable 'newState1' should have a more descriptive name, such as 'stateSnapshot', to better reflect its purpose.